sentry-opentelemetry/sentry-opentelemetry-core/src/main/java/io/sentry/opentelemetry/SentrySpanProcessor.java

```diff
@@ -22,14 +98,140 @@ public boolean isStartRequired() {
   }
 
   @Override
-  public void onEnd(ReadableSpan span) {
-    System.out.println(
-        "hello from onEnd" + Thread.currentThread().getId() + Sentry.getCurrentHub().toString());
-    // TODO end
+  public void onEnd(final @NotNull ReadableSpan otelSpan) {
+    if (!hasSentryBeenInitialized()) {
+      return;
+    }
+
+    if (!Instrumenter.OTEL.equals(Sentry.getCurrentHub().getOptions().getInstrumenter())) {
+      return;
+    }
+
+    final @NotNull SpanContext otelSpanContext = otelSpan.getSpanContext();
+    if (!otelSpanContext.isValid()) {
+      return;
+    }
+
+    final @NotNull TraceData traceData = getTraceData(otelSpan);
+    final @Nullable ISpan sentrySpan = spans.remove(traceData.getSpanId());
+
+    if (sentrySpan == null) {
+      return;
+    }
+
+    if (isSentryRequest(otelSpan)) {
+      return;
+    }
+
+    if (sentrySpan instanceof ITransaction) {
```

Comment:
Do they need distinction?

Reply:
We want to add the OTEL attributes to the transaction context. For spans we want to add data entries.